WalletWasabi/WabiSabi/Client/DependencyGraphResolver.cs

```diff
@@ -0,0 +1,201 @@
+using WalletWasabi.WabiSabi.Client.CredentialDependencies;
+using WalletWasabi.WabiSabi.Crypto;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class DependencyGraphResolver
+	{
+		public DependencyGraphResolver(
+			DependencyGraph graph,
+			ZeroCredentialPool zeroAmountCredentialPool,
+			ZeroCredentialPool zeroVsizeCredentialPool)
+		{
+			Graph = graph;
+			var allInEdges = Enum.GetValues<CredentialType>()
+				.SelectMany(type => Enumerable.Concat<RequestNode>(Graph.Reissuances, Graph.Outputs)
+				.SelectMany(node => Graph.EdgeSets[type].InEdges(node)));
+			DependencyTasks = allInEdges.ToDictionary(edge => edge, _ => new TaskCompletionSource<Credential>(TaskCreationOptions.RunContinuationsAsynchronously));
+
+			ZeroAmountCredentialPool = zeroAmountCredentialPool;
+			ZeroVsizeCredentialPool = zeroVsizeCredentialPool;
+		}
+
+		private DependencyGraph Graph { get; }
+		public ZeroCredentialPool ZeroAmountCredentialPool { get; }
+		public ZeroCredentialPool ZeroVsizeCredentialPool { get; }
+		private Dictionary<CredentialDependency, TaskCompletionSource<Credential>> DependencyTasks { get; }
+
+		public async Task StartConfirmConnectionsAsync(IEnumerable<AliceClient> aliceClients, DependencyGraph dependencyGraph, TimeSpan connectionConfirmationTimeout, CancellationToken cancellationToken)
+		{
+			var aliceNodePairs = PairAliceClientAndRequestNodes(aliceClients, Graph);
+
+			List<SmartRequestNode> smartRequestNodes = new();
+			List<Task> connectionConfirmationTasks = new();
+
+			using CancellationTokenSource ctsOnError = new();
+			using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ctsOnError.Token);
+
+			foreach ((var aliceClient, var node) in aliceNodePairs)
+			{
+				var amountEdgeTaskCompSources = Graph.OutEdges(node, CredentialType.Amount).Select(edge => DependencyTasks[edge]);
+				var vsizeEdgeTaskCompSources = Graph.OutEdges(node, CredentialType.Vsize).Select(edge => DependencyTasks[edge]);
+				SmartRequestNode smartRequestNode = new(
+					Enumerable.Empty<Task<Credential>>(),
+					Enumerable.Empty<Task<Credential>>(),
+					amountEdgeTaskCompSources,
+					vsizeEdgeTaskCompSources,
+					ZeroAmountCredentialPool,
+					ZeroVsizeCredentialPool);
+
+				var amountsToRequest = dependencyGraph.OutEdges(node, CredentialType.Amount).Select(e => (long)e.Value);
+				var vsizesToRequest = dependencyGraph.OutEdges(node, CredentialType.Vsize).Select(e => (long)e.Value);
+
+				var task = smartRequestNode
+					.ConfirmConnectionTaskAsync(
+						connectionConfirmationTimeout,
+						aliceClient,
+						amountsToRequest,
+						vsizesToRequest,
+						node.EffectiveValue,
+						node.VsizeRemainingAllocation,
+						linkedCts.Token)
+					.ContinueWith((t) =>
+					{
+						if (t.IsFaulted && t.Exception is { } exception)
+						{
+							// If one task is failing, cancel all the tasks and throw.
+							ctsOnError.Cancel();
+							throw exception;
```

Comment:
I think here we could log the error (only for the first one)

Reply:
we don't handle errors in any meaningful way yet, so just rethrowing is more correct IMO

Reply:
I agree with Yuval. You can do any logging outside.